src/queue/consumers/visa/consumerCallbacks/syncVisaProposal.ts

```diff
@@ -0,0 +1,131 @@
+  user: ProposalUser,
+  proposalPk: number
+) {
+  const userDataSource = container.resolve<UserDataSource>(
+    Tokens.VisaUserDataSource
+  );
+
+  const experimentDataSource = container.resolve<ExperimentDataSource>(
+    Tokens.VisaExperimentDataSource
+  );
+
+  const experimentUserDataSource = container.resolve<ExperimentUserDataSource>(
+    Tokens.VisaExperimentUserDataSource
+  );
+
+  const createdUser = await userDataSource.create(user);
+  const experiment = await experimentDataSource.getByProposalId(proposalPk);
+  if (experiment && createdUser) {
+    await experimentUserDataSource.create({
+      experimentId: experiment.id,
+      userId: createdUser.id,
+    });
+  }
+}
+
+async function deleteMissingUsersFromExperiment(
+  proposalPk: number,
+  allProposalUsers: ProposalUser[]
+) {
+  const experimentDataSource = container.resolve<ExperimentDataSource>(
+    Tokens.VisaExperimentDataSource
+  );
+  const experimentUserDataSource = container.resolve<ExperimentUserDataSource>(
+    Tokens.VisaExperimentUserDataSource
+  );
+
+  const experiment = await experimentDataSource.getByProposalId(proposalPk);
+  if (!experiment) return;
+
+  const experimentUsers =
+    await experimentUserDataSource.getAllUsersByExperimentId(
+      experiment.id.toString()
+    );
+
+  for (const experimentUser of experimentUsers) {
+    if (
+      !allProposalUsers.some((member) => member.email === experimentUser.email)
+    ) {
+      await experimentUserDataSource.delete({
+        experimentId: experiment.id,
+        userId: experimentUser.id,
+      });
+    }
+  }
+}
+export async function syncVisaProposal(
+  proposalWithNewStatus: ValidProposalMessageData
+) {
+  const proposalDatasource = container.resolve<ProposalDataSource>(
+    Tokens.VisaProposalDataSource
+  );
+
+  const experimentDataSource = container.resolve<ExperimentDataSource>(
+    Tokens.VisaExperimentDataSource
+  );
+
+  const instrumentDataSource = container.resolve<InstrumentDataSource>(
+    Tokens.VisaInstrumentDataSource
+  );
+
+  // Create New Proposal
+  let proposal = await proposalDatasource.get(proposalWithNewStatus.proposalPk);
+  if (!proposal) {
+    proposal = await proposalDatasource.create(proposalWithNewStatus);
+  }
+  // Get Instrument
+  for (const proposalInstrument of proposalWithNewStatus.instruments) {
+    let instrument = await instrumentDataSource.get(proposalInstrument.id);
+
+    if (!instrument) {
+      instrument = await instrumentDataSource.create({
+        id: proposalInstrument.id,
+        name: proposalInstrument.shortCode,
+      });
+    }
+    // Assign Instrument to Proposal and create an Experiment
+    await experimentDataSource.create({
+      proposalPk: proposal.id,
+      instrumentId: instrument.id,
+    });
+  }
+
+  const proposersAndCoproposers = [
+    ...(proposalWithNewStatus.proposer ? [proposalWithNewStatus.proposer] : []),
+    ...proposalWithNewStatus.members,
+  ];
+  // // Create new user for the proposer
```

Comment:
I guess this comment section can be removed since the proposer and members are created at line 120.